src/Language/Futhark/Unused.hs

```diff
@@ -0,0 +1,207 @@
+funcsInDec (ModDec (ModBind _ _ (Just (_, Info mp)) mex _ _)) = funcsInModExp mex `M.union` M.map S.singleton mp
+funcsInDec (ModDec (ModBind _ _ Nothing mex _ _)) = funcsInModExp mex
+funcsInDec _ = M.empty
+
+-- current expression, current function being traversed, map of functions seen so far.
+funcsInExp :: ExpBase Info VName -> VName -> VMap -> VMap
+funcsInExp (Parens ex _) n vm = funcsInExp ex n vm
+funcsInExp (QualParens (QualName _ vn, _) ex _) n vm = funcsInExp ex n (M.map (S.insert vn) vm)
+funcsInExp (TupLit exs _) n vm = foldl' (\x y -> funcsInExp y n x) vm exs
+funcsInExp (RecordLit exs _) n vm = foldl' (\x y -> funcsInExp (getFieldExp y) n x) vm exs
+funcsInExp (ArrayLit exs _ _) n vm = foldl' (\x y -> funcsInExp y n x) vm exs
+funcsInExp (Attr _ ex _) n vm = funcsInExp ex n vm
+funcsInExp (Project _ ex _ _) n vm = funcsInExp ex n vm
+funcsInExp (Not ex _) n vm = funcsInExp ex n vm
+funcsInExp (Assert ex1 ex2 _ _) n vm = funcsInExp ex2 n (funcsInExp ex1 n vm)
+funcsInExp (Constr _ exs _ _) n vm = foldl' (\x y -> funcsInExp y n x) vm exs
+funcsInExp (Update ex1 _ ex2 _) n vm = funcsInExp ex2 n (funcsInExp ex1 n vm)
+funcsInExp (RecordUpdate ex1 _ ex2 _ _) n vm = funcsInExp ex2 n (funcsInExp ex1 n vm)
+funcsInExp (Lambda _ ex _ _ _) n vm = funcsInExp ex n vm
+funcsInExp (OpSection (QualName _ vn) _ _) n vm = M.adjust (S.insert vn) n vm
+funcsInExp (OpSectionLeft (QualName _ vn) _ ex _ _ _) n vm = funcsInExp ex n $ M.adjust (S.insert vn) n vm
+funcsInExp (OpSectionRight (QualName _ vn) _ ex _ _ _) n vm = funcsInExp ex n $ M.adjust (S.insert vn) n vm
+funcsInExp (Ascript ex _ _) n vm = funcsInExp ex n vm
+funcsInExp (Var (QualName _ vn) _ _) n vm = M.adjust (S.insert vn) n vm
+funcsInExp (AppExp app _) n vm =
+  case app of
+    Apply ex1 lst _ ->
+      foldl' (\mp (_, ex) -> funcsInExp ex n mp) (funcsInExp ex1 n vm) lst
+    Coerce ex _ _ -> funcsInExp ex n vm
+    Range ex1 mb_ex inc_ex _ -> foldl' (\x y -> funcsInExp y n x) vm (ex1 : fromInc inc_ex : maybeToList mb_ex)
+    LetPat _ _ ex1 ex2 _ -> funcsInExp ex2 n (funcsInExp ex1 n vm)
+    LetFun vn (_, _, _, _, ex1) ex2 _ ->
+      funcsInExp ex2 n (funcsInExp ex1 vn (M.insert vn S.empty vm)) -- Important case! function defn
+    If ex1 ex2 ex3 _ -> funcsInExp ex3 n $ funcsInExp ex2 n $ funcsInExp ex1 n vm
+    DoLoop _ _ ex1 loop_ex ex2 _ -> funcsInExp (fromLoop loop_ex) n $ funcsInExp ex2 n $ funcsInExp ex1 n vm
+    BinOp (QualName _ vn, _) _ (ex1, _) (ex2, _) _ -> funcsInExp ex2 n $ funcsInExp ex1 n $ M.adjust (S.insert vn) n vm
+    LetWith _ _ sl ex1 ex2 _ -> foldl' (\x y -> funcsInExp y n x) vm $ fromSlice sl <> [ex1, ex2]
+    Index ex sl _ -> foldl' (\x y -> funcsInExp y n x) vm $ ex : fromSlice sl
+    Match ex cases _ -> foldl' (\x y -> funcsInExp y n x) vm $ ex : map fromCase (NE.toList cases)
+funcsInExp _ _ vm = vm
+
+-- funcs inside a module expression. also tracks functions that have different VNames inside and outside the module.
+-- ModVar omitted
```

Comment:
Add a case for `ModVar`.  Wildcard cases are dangerous because they mean we might not notice this function if we ever add more constructors to `ModExp`.